provisioner/aws-manager.js

```diff
@@ -166,6 +271,254 @@ AwsManager.prototype._filterSpotRequests = function (spotReqs) {
+    before we give up on trying to figure out why the instance terminated.
+
+    http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_StateReason.html
+  */
+
+  function plotInstanceDeath (instance, time) {
+    // Let's track when the instance is shut down
+    that.reportInstanceTerminated({
+      provisionerId: that.provisionerId,
+      region: instance.Region,
+      az: instance.Placement.AvailabilityZone,
+      instanceType: instance.InstanceType,
+      workerType: instance.WorkerType,
+      id: instance.InstanceId,
+      spotRequestId: instance.SpotInstanceRequestId,
+      time: dateForInflux(time),
+      launchTime: dateForInflux(instance.LaunchTime),
+      stateCode: instance.State.Code,
+      stateMsg: instance.State.Name,
+      stateChangeCode: instance.StateReason.Code,
+      stateChangeMsg: instance.StateReason.Message,
+    });
+
+    if (instance.StateReason.Code === 'Server.SpotInstanceTermination') {
+      debug('We have a spot price floor!');
+      // Let's figure out what we set the price to;
+      var price;
+
+      deadState.requests.forEach(function (request) {
+        if (!price && request.SpotInstanceRequestId === instance.SpotInstanceRequestId) {
+          price = parseFloat(request.SpotPrice, 10);
+        }
+      });
+
+      apiState.requests.forEach(function (request) {
+        if (!price && request.SpotInstanceRequestId === instance.SpotInstanceRequestId) {
+          price = parseFloat(request.SpotPrice, 10);
+        }
+      });
+
+      if (price) {
+        that.reportSpotPriceFloorFound({
+          region: instance.Region,
+          az: instance.Placement.AvailabilityZone,
+          instanceType: instance.InstanceType,
+          time: new Date().toISOString(),
+          price: price,
+          reason: 'instance-spot-killed',
+        });
+      } else {
+        debug('Could not find a price for a spot-price killed instance');
+      }
+    }
+  }
+
+  // Let's handle instance which already have a state reason, or
+  // save them for a future iteration if not
+  differences.instances.forEach(function (instance) {
+    // Using StateReason instead of StateTransitionReason
+    if (instance.StateReason && instance.StateReason.Code) {
+      debug('found a terminated instance which has a termination reason');
+      plotInstanceDeath(instance, new Date().toISOString());
+    } else {
+      debug('found a terminated instance which awaits a termination reason');
+      that.__awaitingStateReason.push({
+        id: instance.InstanceId,
+        time: new Date().toISOString(),
+        iterationCount: 0,
+      });
+    }
+  });
+
+  // Now, let's try to account for those instance which are awaiting a state reason
+  this.__awaitingStateReason = this.__awaitingStateReason.filter(function (instanceAwaiting) {
+    var keepItInTheList = true;
+    deadState.instances.forEach(function (instanceMightHave) {
+      if (instanceMightHave.InstanceId === instanceAwaiting.id) {
```

Comment:
Fyi, conditions like this can also be done as:
```js
deadState.instances.filter(i => i.InstanceId === instanceAwaiting.id)
                   .forEach(instanceMightHave => {
   ...
})
```
Anyways, it's a matter of taste... Just wanted to give an idea...